feat(AutoSave): create RFC (#3186)
toptal/picasso: docs/decisions/11-autosave.md

# RFC Template

## Problem

Developers want to have auto-save functionality for `RichTextEditor`, and `picasso-forms` doesn't have a built-in solution for this mechanism.

## Proposal

The auto-save feature is fully implemented in `picasso-forms` for all field types.

A mechanism that notifies the parent component about the form values changes will be provided. Then, the parent component will decide how to save this data (in BE or in external storage like local storage) and when. Validation errors will be ignored, for now, so the parent component can decide when to save the data with or without validation errors.

The implementation of the feature can be split into three parts as follows:

- To listen for form state changes, a `final-form` decorator that triggers a callback to notify the parent component with the latest values will be implemented.

>   For performance reasons, there will be made as a subscription functionality so that a component can listen to only specific field value changes.

Reviewer: ```suggestion
- For performance reasons, the decorator will have a subscription functionality so that a component can listen to only specific field value changes.
```

# RFC Template

## Problem

Developers want to have auto-save functionality for `RichTextEditor`, and `picasso-forms` doesn't have a built-in solution for this mechanism.

## Proposal

The auto-save feature is fully implemented in `picasso-forms` for all field types.

A mechanism that notifies the parent component about the form values changes will be provided. Then, the parent component will decide how to save this data (in BE or in external storage like local storage) and when. Validation errors will be ignored, for now, so the parent component can decide when to save the data with or without validation errors.

> The implementation of the feature can be split into three parts as follows:

Reviewer 2: ```suggestion
The implementation of the feature can be split into five parts as follows:
```

- To be able to listen form state changes, we will implement a decorator for `final-form` and trigger a callback to notify parent component with latest values. For performance, we will implement a subscription functionality similar to `final-form` so user can listen only specific fields' value changes.

- We will provide a utility hook from Picasso-forms to provide debouncing (for rapidly changing form values) and visual updates for the parent component ("saving|saved" states).

- There will also be a Picasso component as indicator for showing valuable info for the user about "saving|saved" state.

### Drawbacks and limitations

- It would be a good idea to test form rendering performance since it will start using a new decorator.

## Alternatives

> - Fully implement RichTextEditor auto-save feature on product side without touching Picasso-forms.

Reviewer 3: I'm curious if we were to go this route, it seems like Picasso already exposes enough final-form functionality by exposing the decorators prop for the form.

Looking at the user requests for autosaving in RichTextEditor, they need it for one specific purpose (as far as I remember) - take meeting notes that need to be autosaved.

Not trying to attack the RFC, just being a devil's advocate again :smile: 
What if we don't need this entire code in Picasso if the use case is one-off?

Author: Thanks for asking these questions 🙂 I'm totally up for discussing the need.
I believe that while introducing one indicator component for RichTextEditor with `FormSpy`, we can even take care of the whole form and make it more usable in the future. Also, the whole implementation won't take a huge time I believe. Since, it is very simple IMO

Let's discuss this and decide how to continue after the daily

Author: To be documented here, we did the discussion and decided to extend the functionality to make the feature work for all types of fields.

Also, @mkrl made a great point about using `decorators` prop instead of changing the Form component. I also updated the RFC with this decision 🙇

# RFC Template

## Problem

> Users want to have auto-save functionality for `RichTextEditor` and `picasso-forms` doesn't have a built-in solution for this mechanism.

Reviewer: ```suggestion
Users want to have auto-save functionality for `RichTextEditor`, and `picasso-forms` doesn't have a built-in solution for this mechanism.
```

Adding a comma so that it doesn't say that users want the functionality for  both RichTextEditor and picasso-forms.

## Proposal

> We will fully implement auto-save feature in `picasso-forms` for all field types.

Reviewer: ```suggestion
The auto-save feature is fully implemented in `picasso-forms` for all field types.
```

The `we` from today is different from the `we` of the future + the subject should be the auto-save feature.

> We will provide a mechanism to notify parent component about form values changes and parent component will decide how to save this data (in BE or an external storage like local storage) and when. Validation errors will be ignored for now, so parent component can decide when to save the data with or without validation errors.

Reviewer 2: I think BE has a major drawback when user won't have an internet connection and whole progress won't be saved. If we are going to use BE nevertheless, we should still consider this scenario and have a temporary storage like local storage in that case. Which means that we **_can't avoid local storage_** in any scenario.

Author: As a library, we shouldn't care much about how the data is saved. I mean, IMO, it is not different than a form submission. We don't care about how the user saving the data after form submission. We should only expose this feature like a callback.

Reviewer 2: makes sense. Thank you :bow:

> We will provide a mechanism to notify parent component about form values changes and parent component will decide how to save this data (in BE or an external storage like local storage) and when. Validation errors will be ignored for now, so parent component can decide when to save the data with or without validation errors.

Reviewer: ```suggestion
A mechanism that notifies the parent component about the form values changes will be provided. Then, the parent component will decide how to save this data (in BE or in external storage like local storage) and when. Validation errors will be ignored, for now, so the parent component can decide when to save the data with or without validation errors.
```

> The implementation of the feature can be split to three parts as follows:

Reviewer: ```suggestion
The implementation of the feature can be split into three parts as follows:
```

> - To be able to listen form state changes, we will implement a decorator for `final-form` and trigger a callback to notify parent component with latest values. For performance, we will implement a subscription functionality similar to `final-form` so user can listen only specific fields' value changes.

Reviewer: ```suggestion
- To listen for form state changes, a `final-form` decorator that triggers a callback to notify the parent component with the latest values will be implemented.
For performance reasons, this will be made as a subscription functionality so that a component can listen to only specific field value changes.
```

> - There won't be any changes on `picasso-forms/Form` component since decorator can be plugged in by using `decorators` prop

Reviewer: ```suggestion
- There won't be any changes to the `picasso-forms/Form` component since the decorator can be plugged in by using the `decorators` prop
```

Reviewer: Missing `.` at the end.

> - We will provide a utility hook from `picasso-forms` to provide debouncing (for rapidly changing form values) and visual updates for the parent component ("saving|saved" states).

Reviewer: ```suggestion
- A utility hook from `picasso-forms` will be implemented, to provide debouncing (for rapidly changing form values) and visual updates for the parent component ("saving|saved" states).
```

> - We will also create a Picasso component as indicator for showing valuable info for the user about "saving|saved" state.

Reviewer: ```suggestion
- A Picasso component as an indicator for showing valuable information to the user about the "saving|saved" state.
```

Author: Yeah, I am having the same trouble while writing this kind of text. It should mention about 'end user' specifically. I will fix 👍

Reviewer: Ok. Please have a look through the whole text and rephrase all occurrences adequately because you mentioned `user` at least a couple of times.

### Drawbacks and limitations

> - It would be a good idea to test form rendering performance since it will start using a new decorator.

Reviewer: I don't think that `a good idea to test` is a drawback or limitation. :smile: 
```suggestion
- Form rendering performance might decrease as a result of using a new decorator.
```

## Alternatives

> - Fully implement `RichTextEditor` auto-save feature on product side without touching `picasso-forms`.
> 
>   - While keeping form state, we thought that it would be easier to implement this feature inside Picasso.
> 
> - Implement auto-save feature just for `RichTextEditor` field inside `picasso-forms`.
>   - Instead of implementing the feature just for `RichTextEditor`, we decided to cover all types of input fields so the feature would become more useful for all product teams.

Reviewer: ```suggestion
- Fully implement `RichTextEditor` auto-save feature on the product side without touching `picasso-forms`.
- Implement the auto-save feature just for the `RichTextEditor` field inside `picasso-forms`.
```

I think that explaining what we're doing instead of the alternative is unnecessary here. Hence, the simplified suggestion.
